start adding part one content (#12035)
flutter/website: src/content/tutorial/ui/5-user-input.md

            padding: const EdgeInsets.all(8.0),
            child: TextField(
              maxLength: 5,
              decoration: InputDecoration(
                border: OutlineInputBorder(
                  borderRadius: BorderRadius.all(Radius.circular(35)),
                ),
              ),
            ),
          ),
        ),
      ],
    );
  }
}
```

You have seen some of these widgets in previous lessons: `Row` and
`Padding`. New, though, is the [`Expanded`][] widget. When a child of
a `Row` (or `Column`) is wrapped in `Expanded`, it tells that child to
fill all the available space along the main axis (horizontal for
`Row`, vertical for `Column`) that hasn't been taken by other
children. This makes the `TextField` stretch to take up all the space
*except* what’s taken by other widgets in the row. 

:::tip Tip
`Expanded` is often the solution to "[unbound width/height][]" exceptions. 
:::

Also new is the `TextField` widget. This widget is the star of the
show in this lesson.  This is the basic Flutter widget for text input. 

Thus far, `TextField` has the following configurations. 

* It’s decorated with a rounded border. Notice that the decoration

>   configuration is very similar to how a `Container` and box
>   decorations are decorated.  

Reviewer: ```suggestion
  configuration is very similar to how a `Container` and boxes are decorated.  
```

> * Its `maxLength` property is set to 5, which is how many letters are
>   allowed in any guess in the game.

Reviewer: ```suggestion
* Its `maxLength` property is set to 5 because the game only
  allows guesses of 5-letter words.
```

> ## Handling text with `TextEditingController`

Reviewer: ```suggestion
## Handle text with `TextEditingController`
```
We avoid gerunds in headers

      ],
    );
  }
}
```

A `TextEditingController` is used to read, clear, and modify the text
in a `TextField`. To use it, pass it into the `TextField`.

```dart
class GuessInput extends StatelessWidget {
  GuessInput({super.key, required this.onSubmitGuess});

  final void Function(String) onSubmitGuess;

  final TextEditingController _textEditingController = TextEditingController();

  @override
  Widget build(BuildContext context) {
    return Row(
      children: [
        Expanded(
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: TextField(
              maxLength: 5,
              inputDecoration: InputDecoration(
                border: OutlineInputBorder(
                  borderRadius: BorderRadius.all(Radius.circular(35)),
                ),
              ),
              controller: _textEditingController, // NEW
            ),
          ),
        ),
      ],
    );
  }
}
```

Now, when a user inputs text, you can capture it with the
`_textEditingController`, but you'll need to know *when* to capture
it. The simplest way to react to input is by using the
`TextField.onSubmitted` argument. This argument accepts a callback,

> and the callback will be triggered whenever the user presses the

Reviewer: ```suggestion
and the callback is triggered whenever the user presses the
```

`TextField.onSubmitted`.

```dart
class GuessInput extends StatelessWidget {
  GuessInput({super.key, required this.onSubmitGuess});

  final void Function(String) onSubmitGuess;

  final TextEditingController _textEditingController = TextEditingController();

  @override
  Widget build(BuildContext context) {
    return Row(
      children: [
        Expanded(
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: TextField(
              maxLength: 5,
              inputDecoration: InputDecoration(
                border: OutlineInputBorder(
                  borderRadius: BorderRadius.all(Radius.circular(35)),
                ),
              ),
              controller: _textEditingController,
              onSubmitted: (String input) { // NEW
                print(_textEditingController.text); // Temporary
              }
            ),
          ),
        ),
      ],
    );
  }
}
```

In this case, you could print the `input` passed to the `onSubmitted`
callback directly, but a better user experience clears the text
after each guess: You need a `TextEditingController` to
do that. Update the code as follows:

:::note

> In Dart, it’s good practice to use a [wildcard][] to

Reviewer: ```suggestion
In Dart, it’s good practice to use the `_` [wildcard][] to
```

hide the input to a function that’ll never be used. The following
example does so.
:::

```dart
class GuessInput extends StatelessWidget {
  GuessInput({super.key, required this.onSubmitGuess});

  final void Function(String) onSubmitGuess;

  final TextEditingController _textEditingController = TextEditingController();
  
  @override
  Widget build(BuildContext context) {
    return Row(
      children: [
        Expanded(
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: TextField(
              maxLength: 5,
              inputDecoration: InputDecoration(
                border: OutlineInputBorder(
                  borderRadius: BorderRadius.all(Radius.circular(35)),
                ),
              ),
              controller: _textEditingController,
              onSubmitted: (_) { // UPDATED
                print(_textEditingController.text); // Temporary
                _textEditingController.clear(); // NEW
              } 
            ),
          ),
        ),
      ],
    );
  }
}
```

> ## Input focus

Reviewer: ```suggestion
## Gain input focus
```

Replace the `print` statement with a call to that function.

```dart
class GuessInput extends StatelessWidget {
  GuessInput({super.key, required this.onSubmitGuess});

  final void Function(String) onSubmitGuess;

  final TextEditingController _textEditingController = TextEditingController();

  final FocusNode _focusNode = FocusNode();

  @override
  Widget build(BuildContext context) {
    return Row(
      children: [
        Expanded(
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: TextField(
              maxLength: 5,
              inputDecoration: InputDecoration(
                border: OutlineInputBorder(
                  borderRadius: BorderRadius.all(Radius.circular(35)),
                ),
              ),
              controller: _textEditingController,
              autoFocus: true
              focusNode: _focusNode,
              onSubmitted: (String input) {
                onSubmitGuess(_textEditionController.text.trim())
                _textEditingController.clear();
                _focusNode.requestFocus();
              } 
            ),
          ),
        ),
      ],
    );
  }
}
```

:::note 

> The `trim` function ensures that no whitespace entered;

Reviewer: ```suggestion
The `trim` function prevents whitespace from being entered;
```

```dart
class GuessInput extends StatelessWidget {
  GuessInput({super.key, required this.onSubmitGuess});

  final void Function(String) onSubmitGuess;

  final TextEditingController _textEditingController = TextEditingController();

  final FocusNode _focusNode = FocusNode();

  @override
  Widget build(BuildContext context) {
    return Row(
      children: [
        Expanded(
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: TextField(
              maxLength: 5,
              inputDecoration: InputDecoration(
                border: OutlineInputBorder(
                  borderRadius: BorderRadius.all(Radius.circular(35)),
                ),
              ),
              controller: _textEditingController,
              autoFocus: true
              focusNode: _focusNode,
              onSubmitted: (String input) {
                onSubmitGuess(_textEditionController.text.trim())
                _textEditingController.clear();
                _focusNode.requestFocus();
              } 
            ),
          ),
        ),
      ],
    );
  }
}
```

:::note 

> `trim` is used to ensure there’s no additional whitespace
> entered, which would allow the user to to enter a four letter word
> (plus a whitespace).

Reviewer: ```suggestion
The `trim` function ensures that no whitespace entered;
otherwise, the user could enter a four letter word plus a whitespace.
```

:::

The remaining needed functionality needs to be handled in the parent
widget, `GamePage`.  In the `build` method of that class, add the
`GuessInput` widget under the `Row` widgets in the `Column`’s
children.

```dart
class GamePage extends StatelessWidget {
  final Game _game = Game();

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: Column(
        spacing: 5.0,
        children: [
          for (var guess in _game.guesses)
            Row(
              spacing: 5.0,
              children: [
                for (var letter in guess) Tile(letter),
              ],
            ),
          GuessInput(
            onSubmitGuess: (String guess) {
              // TODO, handle guess
              print(guess); // Temporary
            }
          ),
        ],
      ),
    );
  }
}
```

For the moment, this only prints the guess to prove that it’s wired up

> correctly.  Submitting the guess requires us to use the functionality

Reviewer: ```suggestion
correctly.  Submitting the guess requires using the functionality
```

of a `StatefulWidget`, which you’ll do in the [`StatefulWidget` lesson][].

## Buttons

To improve the UX on mobile and reflect well-known UI, there should
also be a button that can submit the guess.

{%- comment %}
TODO(ewindmill) embed video
{%- endcomment %}

There are many button widgets built into Flutter, like [`TextButton`][],
[`ElevatedButton`][], and the button you’ll use now: [`IconButton`][].  All of
these buttons (and many other interaction widgets) require two
arguments (in addition to their optional arguments): 

* A callback function passed to `onPressed`.   

> * A widget that makes up the content of the button (often a `Text` or `Icon`).

Reviewer: ```suggestion
* A widget that makes up the content of the button (often `Text` or an `Icon`).
```

> Add an icon button to the row widget’s children list in the
> `GuessInput` widget, and give it an [`Icon`][] widget to display. The
> `Icon` widget itself requires icon configuration, and in this case the
> `padding` property, which sets the padding betweenthe edge of the
> button and the icon it wraps, is set to zero. This removes the default
> padding and makes the button smaller.

Reviewer: ```suggestion
Add an icon button to the row widget’s children list in the
`GuessInput` widget, and give it an [`Icon`][] widget to display.
The `Icon` widget requires configuration; in this case, the
`padding` property sets the padding between the edge of the
button and the icon it wraps to zero. This removes the default
padding and makes the button smaller.
```

  final void Function(String) onSubmitGuess;

  final TextEditingController _textEditingController = TextEditingController();
  final FocusNode _focusNode = FocusNode();

    @override
  Widget build(BuildContext context) {
    return Row(
      children: [
        Expanded(...),
        IconButton(
          padding: EdgeInsets.zero,
          icon: Icon(Icons.arrow_circle_up),  
        ),
      ],
    );
  }
}
```

The `IconButton.onPressed` callback should look familiar:

```dart
class GuessInput extends StatelessWidget {
  GuessInput({super.key, required this.onSubmitGuess});

  final void Function(String) onSubmitGuess;

  final TextEditingController _textEditingController = TextEditingController();
  final FocusNode _focusNode = FocusNode();

    @override
  Widget build(BuildContext context) {
    return Row(
      children: [
        Expanded(...),
        IconButton(
          padding: EdgeInsets.zero,
          icon: Icon(Icons.arrow_circle_up),
          onPressed: () {
            onSubmitGuess(_textEditingController.text.trim());
            _textEditingController.clear();
            _focusNode.requestFocus();
          },
        ),
      ],
    );
  }
}
```

> This method is doing the same as the `onSubmitted` callback on the `TextField`. 

Reviewer: ```suggestion
This method does the same as the `onSubmitted` callback on the `TextField`. 
```
(which is?)

:::note Challenge - Share "on submitted" logic.

> You may be thinking, "Shouldn’t we abstract these methods into one

Reviewer: ```suggestion
You might be thinking, "Shouldn’t we abstract these methods into one
```

function and pass it to both inputs?" You could, but it wouldn’t be as
clean as you might think because the callbacks to
`IconButton.onPressed` and `TextField.onSubmitted` require different

> arguments. That said, if the logic was more complex, it'd be better to
> do so. 

Reviewer: ```suggestion
arguments. That said, if the logic was more complex, it would be better to
do so. 
```

Really? Why?

Author: Resolved by making the whole thing more concise in a separate commit
